modules/core/src/main/java/org/apache/ignite/internal/causality/IncrementalVersionedValue.java

```diff
@@ -0,0 +1,209 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.causality;
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static java.util.concurrent.CompletableFuture.failedFuture;
+
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Versioned Value flavor that accumulates updates posted by the {@link #update} method and "commits" them when {@link #complete} or
+ * {@link #completeExceptionally} are called or a storage revision is updated (see constructors for details).
+ */
+public class IncrementalVersionedValue<T> extends AbstractVersionedValue<T> {
+    /** Update mutex. */
+    private final Object updateMutex = new Object();
+
+    /**
+     * Token that was used with the most recent {@link #update} call.
+     *
+     * <p>Multi-threaded access is guarded by {@link #updateMutex}.
+     */
+    private long expectedToken = -1;
+
+    /**
+     * Token that was used with the most recent {@link #completeInternal} call.
+     *
+     * <p>Multi-threaded access is guarded by {@link #updateMutex}.
+     */
+    private long lastCompleteToken = -1;
+
+    /**
+     * Future that will be completed after all updates over the value in context of current causality token will be performed. This
+     * {@code updaterFuture} is {@code null} if no updates in context of current causality token have been initiated. See
+     * {@link #update(long, BiFunction)}.
+     *
+     * <p>Multi-threaded access is guarded by {@link #updateMutex}.
+     */
+    private CompletableFuture<T> updaterFuture = completedFuture(getDefault());
+
+    /**
+     * Constructor.
+     *
+     * @param observableRevisionUpdater A closure intended to connect this VersionedValue with a revision updater, that this
+     *         VersionedValue should be able to listen to, for receiving storage revision updates. This closure is called once on a
+     *         construction of this VersionedValue and accepts a {@code Function<Long, CompletableFuture<?>>} that should be called on every
+     *         update of storage revision as a listener.
+     * @param maxHistorySize Size of the history of changes to store, including last applied token.
+     * @param defaultValueSupplier Supplier of the default value, that is used on {@link #update(long, BiFunction)} to evaluate the
+     *         default value if the value is not initialized yet. It is not guaranteed to execute only once.
+     */
+    public IncrementalVersionedValue(
+            @Nullable Consumer<Function<Long, CompletableFuture<?>>> observableRevisionUpdater,
```

Comment:
> Are there benefits of using Function<Long, ...> instead of LongFunction<...>?

There are no benefits, this interface was not created by me and is used in may other components. I despise it and intend to get rid of it as soon as I have time.

> having this value in the constructor looks completely backwards to me

Again, this comes from the previous design, I decided not to change it (yet) in order to reduce the size of this PR because this change is not completely necessary. Anyway, where can we move this subscription?

Reply:
Ok, thank you!